prov/tcp/src/tcpx_ep.c

```diff
@@ -483,10 +483,72 @@ static int tcpx_pep_create_listen_socket(struct tcpx_pep *pep, void *src_addr,
 	return FI_SUCCESS;
 }
 
+static int tcpx_ep_setname(fid_t fid, void *addr, size_t addrlen)
+{
+	struct tcpx_ep *tcpx_ep;
+	struct tcpx_pep *tcpx_pep;
+	uint32_t addr_format;
+
+	switch (addrlen) {
+	case FI_SOCKADDR_IN:
+		addr_format = FI_SOCKADDR_IN;
+		break;
+	case FI_SOCKADDR_IN6:
+		addr_format = FI_SOCKADDR_IN6;
+		break;
+	default:
+		return -FI_EINVAL;
+	}
+
+	switch (fid->fclass) {
+	case FI_CLASS_EP:
+		tcpx_ep = container_of(fid, struct tcpx_ep, util_ep.ep_fid);
+		return (bind(tcpx_ep->conn_fd,
+			     addr, addrlen))?-errno : FI_SUCCESS;
+
+	case FI_CLASS_PEP:
+		tcpx_pep = container_of(fid, struct tcpx_pep, util_pep.pep_fid);
+		if (tcpx_pep->sock != INVALID_SOCKET)
+			return -FI_EINVAL;
+
+		return tcpx_pep_create_listen_socket(tcpx_pep, addr,
+						     addr_format, addrlen);
+	default:
+		FI_WARN(&tcpx_prov, FI_LOG_EP_DATA,"Invalid argument\n");
+		return -FI_EINVAL;
+	}
+	return -FI_EINVAL;
+}
+
+static int tcpx_ep_getname(fid_t fid, void *addr, size_t *addrlen)
+{
+	struct tcpx_ep *tcpx_ep;
+	struct tcpx_pep *tcpx_pep;
+	size_t addrlen_in = *addrlen;
+
+	switch (fid->fclass) {
+	case FI_CLASS_EP:
+		tcpx_ep = container_of(fid, struct tcpx_ep, util_ep.ep_fid);
+		if (getsockname(tcpx_ep->conn_fd, addr,(socklen_t *) addrlen))
+			return (addrlen_in < *addrlen)? -FI_ETOOSMALL: -errno;
+
+		return FI_SUCCESS;
+	case FI_CLASS_PEP:
+		tcpx_pep = container_of(fid, struct tcpx_pep, util_pep.pep_fid);
+		if (getsockname(tcpx_pep->sock, addr, (socklen_t *)addrlen))
+			return (addrlen_in < *addrlen)? -FI_ETOOSMALL: -errno;
+
+		return FI_SUCCESS;
+	default:
+		FI_WARN(&tcpx_prov, FI_LOG_EP_DATA,"Invalid argument\n");
+		return -FI_EINVAL;
+	}
+}
```

Comment:
This would be better just being split into 2 calls -- one for EPs and another for PEPs.